File: 第一阶段/1.8-列表/Lists.cs
// C# 列表操作示例
// 对应文章：1.8 列表与数组

using System;
using System.Collections.Generic;
using System.Linq;

// ==================== 数组（固定大小）====================
Console.WriteLine("=== 数组 vs List ===");
int[] arr1 = { 1, 2, 3 };              // 数组（固定大小）
int[] arr2 = new int[5];               // 默认值为 0
Console.WriteLine($"数组: [{string.Join(", ", arr1)}]");

// List（动态大小，类似 Python 的 list）
List<int> list1 = new() { 1, 2, 3 };
List<int> list2 = new List<int>();     // 空列表
Console.WriteLine($"List: [{string.Join(", ", list1)}]");

// ==================== 添加元素 ====================
Console.WriteLine("\n=== 添加元素 ===");
list1.Add(6);                          // 添加到末尾
list1.Insert(0, 0);                    // 插入到指定位置
list1.AddRange(new[] { 7, 8 });        // 添加多个
Console.WriteLine($"操作后: [{string.Join(", ", list1)}]");

// ==================== 删除元素 ====================
Console.WriteLine("\n=== 删除元素 ===");
list1.Remove(6);                       // 删除第一个匹配项
list1.RemoveAt(0);                     // 删除指定索引
list1.RemoveAll(x => x > 7);          // 删除所有匹配项
Console.WriteLine($"操作后: [{string.Join(", ", list1)}]");

// ==================== 查找 ====================
Console.WriteLine("\n=== 查找 ===");
var nums = new List<int> { 10, 20, 30, 40, 50 };
int index = nums.IndexOf(30);          // 查找索引: 2
bool exists = nums.Contains(30);       // 是否包含: true
Console.WriteLine($"IndexOf(30) = {index}");
Console.WriteLine($"Contains(30) = {exists}");

// ==================== 排序和反转 ====================
Console.WriteLine("\n=== 排序和反转 ===");
var sortList = new List<int> { 3, 1, 4, 1, 5, 9, 2, 6 };
sortList.Sort();                       // 排序（原地）
Console.WriteLine($"Sort(): [{string.Join(", ", sortList)}]");
sortList.Reverse();                    // 反转（原地）
Console.WriteLine($"Reverse(): [{string.Join(", ", sortList)}]");

// 带比较器排序
var words = new List<string> { "banana", "apple", "cherry" };
words.Sort((a, b) => a.Length.CompareTo(b.Length));
Console.WriteLine($"按长度排序: [{string.Join(", ", words)}]");

// ==================== 切片 ====================
Console.WriteLine("\n=== 切片（C# 8.0+ Range）===");
var sliceList = Enumerable.Range(0, 10).ToList();
var sub = sliceList.GetRange(1, 3);     // 传统方式：从索引1开始取3个
var rangeSub = sliceList[1..4];          // C# 8.0+ 方式
Console.WriteLine($"GetRange(1,3): [{string.Join(", ", sub)}]");
Console.WriteLine($"[1..4]:        [{string.Join(", ", rangeSub)}]");

// ==================== LINQ（等价于 Python 列表推导式）====================
Console.WriteLine("\n=== LINQ ===");
var numbers = new List<int> { 1, 2, 3, 4, 5 };
var squares = numbers.Select(x => x * x).ToList();           // 平方
var even = numbers.Where(x => x % 2 == 0).ToList();          // 偶数
int sum = numbers.Sum();
double avg = numbers.Average();
Console.WriteLine($"平方: [{string.Join(", ", squares)}]");
Console.WriteLine($"偶数: [{string.Join(", ", even)}]");
Console.WriteLine($"求和: {sum}, 平均: {avg}");

// ==================== 浅拷贝 vs 深拷贝 ====================
Console.WriteLine("\n=== 浅拷贝 vs 深拷贝 ===");
var original = new List<List<int>> { new() { 1, 2 }, new() { 3, 4 } };
var shallow = new List<List<int>>(original);  // 浅拷贝
shallow[0][0] = 999;
Console.WriteLine($"浅拷贝修改后 original[0][0] = {original[0][0]}");  // 999（被影响！）

var deep = original.Select(x => new List<int>(x)).ToList();  // 深拷贝
deep[0][0] = 1;
Console.WriteLine($"深拷贝修改后 original[0][0] = {original[0][0]}");  // 999（不受影响）

File: 第一阶段/1.8-列表/lists.py
# Python 列表操作示例
# 对应文章：1.8 列表与数组

# ==================== 列表创建 ====================
print("=== 列表创建 ===")
lst1 = [1, 2, 3]
lst2 = list([1, 2, 3])
lst3 = list()  # 空列表
lst4 = []      # 空列表（更常用）

# 初始化
lst5 = [0] * 5       # [0, 0, 0, 0, 0]
lst6 = [None] * 3    # [None, None, None]
print(f"lst1 = {lst1}")
print(f"lst5 = {lst5}")

# ==================== 添加元素 ====================
print("\n=== 添加元素 ===")
lst = [1, 2, 3]
lst.append(6)          # 添加到末尾: [1, 2, 3, 6]
lst.insert(0, 0)       # 插入到指定位置: [0, 1, 2, 3, 6]
lst.extend([7, 8])     # 添加多个: [0, 1, 2, 3, 6, 7, 8]
print(f"操作后: {lst}")

# ==================== 删除元素 ====================
print("\n=== 删除元素 ===")
lst = [1, 2, 3, 2, 4]
lst.remove(2)          # 删除第一个匹配项: [1, 3, 2, 4]
print(f"remove(2): {lst}")

val = lst.pop()        # 删除并返回最后一个
print(f"pop(): {val}, 剩余: {lst}")

val = lst.pop(0)       # 删除并返回指定索引
print(f"pop(0): {val}, 剩余: {lst}")

del lst[0]             # 删除指定索引
print(f"del[0]:  剩余: {lst}")

# lst.clear()           # 清空列表

# ==================== 查找 ====================
print("\n=== 查找 ===")
lst = [10, 20, 30, 40, 50]
index = lst.index(30)   # 查找索引: 2
exists = 30 in lst      # 是否包含: True（运算符，不是方法！）
print(f"index(30) = {index}")
print(f"30 in lst = {exists}")

# ==================== 排序和反转 ====================
print("\n=== 排序和反转 ===")
lst = [3, 1, 4, 1, 5, 9, 2, 6]
sorted_lst = sorted(lst)  # 排序（返回新列表，不修改原列表）
print(f"原列表:    {lst}")
print(f"sorted():  {sorted_lst}")

lst.sort()              # 排序（原地修改）
print(f"sort()后:  {lst}")

lst.reverse()           # 反转（原地修改）
print(f"reverse(): {lst}")

# 按自定义规则排序
words = ["banana", "apple", "cherry"]
by_length = sorted(words, key=len)
print(f"按长度排序: {by_length}")

# ==================== 切片 ====================
print("\n=== 切片 ===")
lst = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
print(f"原列表:  {lst}")
print(f"[1:4]:   {lst[1:4]}")      # [1, 2, 3]
print(f"[:5]:    {lst[:5]}")        # [0, 1, 2, 3, 4]
print(f"[::2]:   {lst[::2]}")       # [0, 2, 4, 6, 8]（每隔一个）
print(f"[::-1]:  {lst[::-1]}")      # [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]（反转）

# ==================== 列表推导式 ====================
print("\n=== 列表推导式 ===")
numbers = [1, 2, 3, 4, 5]
squares = [x * x for x in numbers]                          # [1, 4, 9, 16, 25]
even = [x for x in numbers if x % 2 == 0]                  # [2, 4]
matrix = [[i * j for j in range(1, 4)] for i in range(1, 4)]  # 嵌套推导式
print(f"平方:  {squares}")
print(f"偶数:  {even}")
print(f"矩阵:  {matrix}")

# ==================== 内置函数 ====================
print("\n=== 内置聚合函数 ===")
numbers = [1, 2, 3, 4, 5]
print(f"sum:     {sum(numbers)}")      # 15
print(f"avg:     {sum(numbers) / len(numbers)}")  # 3.0
print(f"max:     {max(numbers)}")      # 5
print(f"min:     {min(numbers)}")      # 1
print(f"len:     {len(numbers)}")      # 5

# ==================== 浅拷贝 vs 深拷贝 ====================
print("\n=== 浅拷贝 vs 深拷贝 ===")
import copy

original = [[1, 2], [3, 4]]

# 浅拷贝：只复制一层
shallow = original.copy()  # 或 list(original) 或 original[:]
shallow[0][0] = 999
print(f"浅拷贝修改后 original: {original}")  # [[999, 2], [3, 4]]（被影响了！）

# 深拷贝：递归复制所有层
original = [[1, 2], [3, 4]]
deep = copy.deepcopy(original)
deep[0][0] = 999
print(f"深拷贝修改后 original: {original}")  # [[1, 2], [3, 4]]（不受影响！）

# ==================== 列表当栈使用 ====================
print("\n=== 列表当栈 ===")
stack = []
stack.append(1)    # push
stack.append(2)
stack.append(3)
print(f"入栈: {stack}")
val = stack.pop()  # pop（LIFO）
print(f"出栈: {val}, 栈: {stack}")

# 注意：列表当队列效率低！pop(0) 是 O(n)，用 collections.deque
from collections import deque
queue = deque([1, 2, 3])
queue.append(4)     # 入队
val = queue.popleft()  # 出队（O(1)）
print(f"队列出队: {val}, 队列: {queue}")
